src/main/java/org/cactoos/iterator/Paged.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2020 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.iterator;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.atomic.AtomicReference;
+import org.cactoos.Func;
+import org.cactoos.Scalar;
+import org.cactoos.func.UncheckedFunc;
+import org.cactoos.scalar.Sticky;
+import org.cactoos.scalar.Unchecked;
+
+/**
+ * Paged iterator.
+ * Elements will continue to be provided so long as {@code next} produces
+ * non-empty iterators.
+ *
+ * <p>There is no thread-safety guarantee.
+ *
+ * @param <I> Subtype of item iterator
+ * @param <X> Type of item
+ * @since 0.49
+ */
+public final class Paged<I extends Iterator<X>, X> implements Iterator<X> {
+
+    /**
+     * Current element.
+     */
+    private final AtomicReference<Unchecked<I>> current;
+
+    /**
+     * Function to get the next element.
+     */
+    private final UncheckedFunc<I, I> subsequent;
+
+    /**
+     * Ctor.
+     * @param first First element.
+     * @param next Function to get the next element.
+     */
+    public Paged(final Scalar<I> first, final Func<I, I> next) {
+        this.current =
+            new AtomicReference<>(
+                new Unchecked<>(
+                    new Sticky<>(first)
+                )
+            );
+        this.subsequent = new UncheckedFunc<>(next);
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!this.current.get().value().hasNext()) {
+            final I next = this.subsequent.apply(
+                this.current.get().value()
+            );
+            this.current.set(
+                new Unchecked<>(
+                    new Sticky<>(() -> next)
+                )
```

Comment:
@baudoliver7 it seems to me that we don't need at all to wrap into a `Scalar`, let's just store `next` itself here

Reply:
@victornoel `next` is just the function that retrieve the next item. The value found is stored in `current` variable that is a `AtomicReference`. We wrap the `Scalar` like this to give us possibility to update its value later while keeping it immutable.

Reply:
@baudoliver7 you can hide mutability behind a scalar, but it is still mutability. We have a tendency to use `AtomicReference` and co in cactoos to avoid non-final fields, but then, if it's mutable, it's mutable, that's life.

Reply:
@victornoel Yeah :)

Reply:
@baudoliver7 we are still missing that change to have everything for merging :)

Reply:
@victornoel In this case, `this.current` should be `non final`. Isn't it ?

Reply:
@baudoliver7 it should be:
```java
private final AtomicReference<I> current;
```

Reply:
@victornoel By this way, constructor will be : 
```
public Paged(final I first, final Func<I, I> next) {
    this.current = new AtomicReference<I>(first);
    this.subsequent = next;
}
```
Isn't it ?

Reply:
@baudoliver7 ha,  hmmm, looks like I didn't think this through... but then yes! I always thought it was strange to have a `Scalar` as first... an `Iterator` is already an object with behaviour by itself, no need for `Scalar` which is useful when you cannot have your own implementation of types such as `String` or `Integer`. You can of course also add a todo if you are sick of all those extra changes I am adding, this is what they are for :)

And now I really wonder why `iterable.Paged` is taking `Iterator` and not `Iterable`... ! Could you also add a todo in `iterable.Paged`: "Replace the constructor so that instead it takes a `Iterable` as `first` and a `Func<Iterable, Iterable>` as `next` in order to homogenise it with the other `Iterable` implementations." Thx!

Reply:
@victornoel I done all extra changes about using `I` directly and I reported `Replace the constructor...` as a todo.